torch_xla/runtime.py

```diff
@@ -12,10 +12,41 @@
 import torch_xla.utils.utils as xu
 import torch_xla._internal.tpu as tpu
 from torch_xla.experimental import plugins
+from torch_xla import runtime
 
 R = TypeVar('R')
 FN = TypeVar('FN')
 
+# Note [Dynamo WORLD_SIEZ and ORDINAL]
+# Belows are workaround to cache the ordinal and world_size such that
+# Dynamo won't do graph breaks when runtime.xrt_world_size() and runtime.global_ordinal() are called.
+_WORLD_SIZE = None
+_ORDINAL = None
+
+
+def _init_world_size_ordinal():
+  global _WORLD_SIZE, _ORDINAL
+
+  # Dynamo doesn't support XRT or multithreaded runtime. See Note [V3-8 Threading]
+  if not runtime.using_pjrt() or runtime.addressable_device_count() > 1:
+    return
+
+  if _WORLD_SIZE is None:
+    _WORLD_SIZE = runtime.world_size()
+    _ORDINAL = runtime.global_ordinal()
+
+
+def run_once(func):
```

Comment:
What is this `run_once` for? It's a neat idea, but I see you opted for global variables for world size and ordinal

Reply:
Without the `@run_once` for `using_pjrt`, the test `test_mp_replication` will fail with build the dynamic graph for dynamo compile.

Reply:
Do you know why that is? Is it because now all calls to get e.g. `world_size` go through functions wrapped in `requires_pjrt`, which in turn actually is checking an env var (`device_type` and `_maybe_select_default_device`)? Whereas before the call would have been stopped by `xm.world_size`.

Reply:
Check my reply here: https://github.com/pytorch/xla/pull/7679#issuecomment-2246520735